include/envoy/upstream/cluster_manager.h

```diff
@@ -147,6 +170,17 @@ class ClusterManager {
    * @return std::string the local cluster name, or "" if no local cluster was configured.
    */
   virtual const std::string& localClusterName() const PURE;
+
+  /**
+   * These two methods allow to register and unregister callbacks for cluster lifecycle events
+   * in the ClusterManager.
+   * The callbacks will be registered in a thread local slot and the callbacks will be executed
+   * on the thread that registered them.
+   *
+   * @param callbacks are the ClusterUpdateCallbacks to add or remove to the cluster manager.
+   */
+  virtual void addClusterUpdateCallbacks(ClusterUpdateCallbacks& callbacks) PURE;
```

Comment:
I still would recommend naming these `addThreadLocal...` and `removeThreadLocal...`. Either way, I would clarify in the method docs that the caller needs to call these functions on every thread in which they expect to receive callbacks. It's not completely clear IMO that is what happens.

Reply:
Done.